src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -33,140 +32,192 @@ internal partial class SyntacticClassificationTaggerProvider
                 TimeSpan diffTimeout)
+                : base(taggerProvider.ThreadingContext, assertIsForeground: false)
             {
+                _taggerProvider = taggerProvider;
                 _subjectBuffer = subjectBuffer;
-                _notificationService = notificationService;
                 _listener = asyncListener;
                 _typeMap = typeMap;
-                _taggerProvider = taggerProvider;
                 _diffTimeout = diffTimeout;
-                _workQueue = new AsynchronousSerialWorkQueue(taggerProvider._threadingContext, asyncListener);
-                _reportChangeCancellationSource = new CancellationTokenSource();
+                _workQueue = new AsyncBatchingWorkQueue<ITextSnapshot>(
+                    TimeSpan.FromMilliseconds(TaggerConstants.NearImmediateDelay),
+                    ProcessChangesAsync,
+                    equalityComparer: null,
+                    asyncListener,
+                    _disposalCancellationSource.Token);
 
-                _lastLineCache = new LastLineCache();
+                _lastLineCache = new LastLineCache(taggerProvider.ThreadingContext);
 
                 _workspaceRegistration = Workspace.GetWorkspaceRegistration(subjectBuffer.AsTextContainer());
                 _workspaceRegistration.WorkspaceChanged += OnWorkspaceRegistrationChanged;
 
                 if (_workspaceRegistration.Workspace != null)
-                {
                     ConnectToWorkspace(_workspaceRegistration.Workspace);
+
+                // Prefer to hear about text changes in the BG so we can use as little 
+                if (_subjectBuffer is ITextBuffer2 textBuffer2)
+                {
+                    textBuffer2.ChangedOnBackground += this.OnSubjectBufferChanged;
+                }
+                else
+                {
+                    _subjectBuffer.Changed += this.OnSubjectBufferChanged;
                 }
             }
 
-            private void OnWorkspaceRegistrationChanged(object sender, EventArgs e)
+            public event EventHandler<SnapshotSpanEventArgs>? TagsChanged;
+
+            private IClassificationService? TryGetClassificationService(ITextSnapshot snapshot)
+            {
+                if (_workspace is not { } workspace)
+                    return null;
+
+                var languageServices = workspace.Services.GetLanguageServices(snapshot.ContentType);
+                if (languageServices == null)
+                    return null;
+
+                return languageServices.GetService<IClassificationService>();
+            }
+
+            #region Workspace Hookup
+
+            private void OnWorkspaceRegistrationChanged(object? sender, EventArgs e)
+            {
+                var token = _listener.BeginAsyncOperation(nameof(OnWorkspaceRegistrationChanged));
+                var task = SwitchToMainThreadAndHookupWorkspaceAsync();
+                task.CompletesAsyncOperation(token);
+            }
```

Comment:
i movd workspace registration change to the UI thread.  I couldn't reason at all about the threadsafety of things (like _workspace) if they can change on arbitrary threads.